rag/llm/key_utils.py

```
@@ -31,4 +33,71 @@ def _normalize_replicate_key(key):
+    if missing).
+    """
+    if isinstance(key, dict):
+        payload = key
+    elif isinstance(key, str):
+        try:
+            payload = json.loads(key)
+        except (json.JSONDecodeError, TypeError):
+            raise ModelException(
+                "Google Vertex AI key is not valid JSON. Expected an object with 'google_project_id', 'google_region', and 'google_service_account_key' (see api/apps/llm_app.py:291 for the schema).",
+                retryable=False,
+            )
```

Ruff (0.15.21)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)